packages/wordcount/src/index.js

```diff
@@ -50,56 +62,62 @@ function loadSettings( type, userSettings ) {
+ * Count the words in text
  *
  * @param {string} text     The text being processed
- * @param {string} regex    The regular expression pattern being matched
- * @param {Object} settings Settings object containing regular expressions for each strip function
+ * @param {RegExp|undefined} regex    The regular expression pattern being matched
+ * @param {WordCountSettings} settings Settings object containing regular expressions for each strip function
  *
- * @return {Array|{index: number, input: string}} The matched string.
+ * @return {number} The matched string.
  */
-function matchWords( text, regex, settings ) {
+function countWords( text, regex, settings ) {
+	if ( typeof regex === 'undefined' ) {
+		return 0;
+	}
 	text = flow(
-		stripTags.bind( this, settings ),
-		stripHTMLComments.bind( this, settings ),
-		stripShortcodes.bind( this, settings ),
-		stripSpaces.bind( this, settings ),
-		stripHTMLEntities.bind( this, settings ),
-		stripConnectors.bind( this, settings ),
-		stripRemovables.bind( this, settings )
+		stripTags.bind( null, settings ),
+		stripHTMLComments.bind( null, settings ),
+		stripShortcodes.bind( null, settings ),
+		stripSpaces.bind( null, settings ),
+		stripHTMLEntities.bind( null, settings ),
+		stripConnectors.bind( null, settings ),
+		stripRemovables.bind( null, settings )
 	)( text );
 	text = text + '\n';
-	return text.match( regex );
+	return text.match( regex )?.length ?? 0;
 }
 
 /**
- * Match the regex for either 'characters_excluding_spaces' or 'characters_including_spaces'
+ * Count the characters in text
  *
  * @param {string} text     The text being processed
- * @param {string} regex    The regular expression pattern being matched
- * @param {Object} settings Settings object containing regular expressions for each strip function
+ * @param {RegExp|undefined} regex    The regular expression pattern being matched
+ * @param {WordCountSettings} settings Settings object containing regular expressions for each strip function
  *
- * @return {Array|{index: number, input: string}} The matched string.
+ * @return {number} Count of characters.
  */
-function matchCharacters( text, regex, settings ) {
+function countCharacters( text, regex, settings ) {
+	if ( typeof regex === 'undefined' ) {
+		return 0;
+	}
 	text = flow(
-		stripTags.bind( this, settings ),
-		stripHTMLComments.bind( this, settings ),
-		stripShortcodes.bind( this, settings ),
-		stripSpaces.bind( this, settings ),
-		transposeAstralsToCountableChar.bind( this, settings ),
-		transposeHTMLEntitiesToCountableChars.bind( this, settings )
+		stripTags.bind( null, settings ),
+		stripHTMLComments.bind( null, settings ),
+		stripShortcodes.bind( null, settings ),
+		transposeAstralsToCountableChar.bind( null, settings ),
+		stripSpaces.bind( null, settings ),
+		transposeHTMLEntitiesToCountableChars.bind( null, settings )
 	)( text );
 	text = text + '\n';
-	return text.match( regex );
+	return text.match( regex )?.length ?? 0;
```

Comment:
Is it possible that `String#match` would result in a valid match object (with `length`) where `length` would be non-numeric?

Reply:
Yep! I came across this when typing `autop` (thanks to TypeScript types, I had no idea). In that case, I added a type assertion:

https://github.com/WordPress/gutenberg/blob/a67c9a9c69d8daee389b5d89ef460db7718b90e8/packages/autop/src/index.js#L63-L67

> Return value
> An `Array` whose contents depend on the presence or absence of the global (`g`) flag, or `null` if no matches are found.
> - If the `g` flag is used, all results matching the complete regular expression will be returned, but capturing groups will not.
> - if the `g` flag is not used, only the first complete match and its related capturing groups are returned. In this case, the returned item will have additional properties as described below.

[from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match#Return_value)

In a nutshell:

```js
'foo'.match(/f/)
// ["f", index: 0, input: "foo", groups: undefined]
'foo'.match(/f/g) // match behaves differently with "g" flag
// ["f"]
```

It's possible to inspect the flags. With a configurable regex like this, we might throw and error or at least show a warning if a regex with the `g` flag is provided:

```js
if ( regex.flags.includes( 'g' ) ) {
  throw new Error( '💥');
}

Reply:
@aduth [docs say](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match) it returns `Array` or `null`.

Reply:
I guess the question for me is: Is the following not enough if it's an `Array` or `null`?

```suggestion
	return text.match( regex )?.length;
```

Reply:
@aduth in this case wouldn't the function return `undefined` and not `0`? We want it to return `number`.

Reply:
Oh! This is yet another case where I eagerly look to the end of the chain of an optional chaining expression. I was considering the `??` to be operating exclusively on `length`, not considering that it's also accounting for the `?.` short-circuit.

Funny enough, this is the exact sort of "gotcha" I just finished documenting in #22029 😆 

>This is an easy trap to fall into when eagerly assuming that a result of `trim()` will always return a string value, overlooking the fact the optional chaining may have caused evaluation to abort earlier with a value of `undefined`.